### ISSUE TITLE
Suppport soname for cdylibs

### ISSUE THREAD

**Issue by kornelski** [rust-lang/cargo#5045]
`rustc` supports setting `soname` via [`-C link-arg`](https://github.com/rust-lang/rust/issues/37529#issuecomment-301603498). However, it [can't be used from Cargo](https://github.com/rust-lang/rust/issues/43190#issuecomment-314871004), since `RUSTFLAGS` affects all invocations of the compiler, and it could potentially set `soname` on dependencies.

Some previous discussions about `soname` support were dismissed by the fact that Rust's native ABI is not stable. That is not relevant for cdylibs, because the C ABI exposed by cdylibs *is* stable. 

I'm interested in using this feature in [`cargo-deb`](https://github.com/mmstick/cargo-deb). `soname` is required to build first-class Linux libraries ([e.g. for Debian](https://lintian.debian.org/tags/sharedobject-in-library-directory-missing-soname.html)) with Rust.

I think it'd be wonderful if Cargo automatically set `soname` of `cdylib` to semver-major when version is >= 1.0.0 (i.e. don't set any for 0.x.x), since soname and semver-major promise basically the same thing. However, a separate Cargo.toml field or env var would work as well.

**Comment by kornelski** [comment]
The lack of support for linker flags is one of issues preventing librsvg from switching to Cargo. It needs flags like this:

```
	-version-info @RSVG_LT_VERSION_INFO@ \
	-export-dynamic \
	-no-undefined \
	-export-symbols-regex "^rsvg_.*" 
```

**Comment by sdroege** [comment]
Those are all libtool flags and not direct linker flags that `ld` or `link.exe` would handle.

**Comment by kornelski** [comment]
Ah, ok. Still needs post build script or something.

**Comment by lu-zero** [comment]
@kornelski I'm updating my patch, then we can work on the installing phase together if you like :)

I need it already for crav1e and the relibc developer seem interested as well.

**Comment by federicomenaquintero** [comment]
From !6298, am I correct in understanding that the remaining work is to figure out platform-specific linker arguments to define the soname and such?

**Comment by lu-zero** [comment]
The [install phase rfc](https://github.com/rust-lang/rfcs/pull/2376) and [this](https://github.com/rust-lang/crates-io-cargo-teams/issues/14) have more details on what is missing.

I might take this weekend to publish my draft applet so we can experiment a bit outside cargo if you want to join me in this :)

**Comment by kornelski** [comment]
But this is about `cargo build`, not `cargo install`, so the RFC won't help.

**Comment by lu-zero** [comment]
What I'm preparing is something along the lines of:
```
cargo c install --prefix=/usr --destdir=/staging/place
```

**Comment by kornelski** [comment]
But isn't that a workaround? It reminds me of ugly hacks needed for `make install`. The products are already in `target/`, they're just built wrong.

**Comment by lu-zero** [comment]
It is not. 
If you want to build only you just issue a `build` command. 

As a packager I enjoy having a working install target that gets all I need in the right places without having to reinvent the wheel every time. Nobody is forcing you to use what I'm writing :)

**Comment by kornelski** [comment]
Oh, you mean you're writing "`cargo-c`"? I assumed that `c` was a typo, and you were suggesting `cargo install` as the workaround (and that RFC goes way beyond staging build products, into making `cargo install` a package manger that can't uninstall what it has installed, so I'm very dismayed by it).

**Comment by lu-zero** [comment]
yes, I try to make an applet to generate all that's needed (.pc files, .h files, etc) and install it without to have to repeat myself too much (currently we are having a small vlc workshop so I might or might not get to publish it this weekend)

**Comment by lu-zero** [comment]
Now I have two ways to help on this:
- https://crates.io/crates/cdylib-link-lines if you want just to use `build.rs` and just build the .so
- https://github.com/lu-zero/cargo-c if you want to build and install a C-lib-like package.

Ideally I could make a separate crate for the `.pc` creation as well if somebody needs it.

**Comment by h0ru5** [comment]
new to rust/cargo: would there be a possibility to write a plugin that does set the soname as post-production step? (yes, it is a workaround until mainline can do it).

Right now I try to do the same on the side consuming the .so

**Comment by lu-zero** [comment]
it is, you may using cdylib-link-lines (even if it could enjoy an overhaul).

**Comment by MartinKolbAtWork** [comment]
I am facing exactly this issue when trying to use **Rust-based libraries** (i.e. I used the two Rust-based WebAssembly engines Wasmtime and Wasmer) in a **C/C++ code-base**.

These Rust libraries have **no SONAMES**, which causes the **CMake build** to produce **wrong binaries**.

I opened an issue at CMake, but these folks clearly confirmed that shared libraries **without an SONAME** entry are a **bug** that needs to be fixed by **the one building the library**:
https://gitlab.kitware.com/cmake/cmake/-/issues/22307#note_971562

Could you please continue working on this topic? Otherwise consuming Rust-based libraries with the popular CMake build system will remain awkward.

**Comment by lu-zero** [comment]
You can use [cargo-c](https://crates.io/crates/cargo-c) for your use-case I think.

**Comment by stephan57160** [comment]
Interested in the same, with also an Android context.

Because of Android support, we're also already using cargo-ndk.
As a workaround, how to use cargo-ndk and cargo-c together ? Any idea would be appreciated.

**Comment by lu-zero** [comment]
Ideally you might get around setting the `CARGO` to `cargo capi` if cargo-ndk calls normally `cargo build` `cargo install` or extend a little cargo-ndk to know about cargo-c and call it instead of cargo if present.

For android you shouldn't have problems not using cargo-c for it.

**Comment by weihanglo** [comment]
Out of curious. What's the missing part to make progress on this issue?

From my understanding, [`rustc-link-arg` family](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-link-arg) in build script solve this issue somehow. We could also simply run something like the following to build a shared library with SONAME:

``` 
cargo rustc --crate-type cdylib -- -C link-arg=-Wl,-soname,libmycrate.so
```

Sorry I am a bit lost on this thread…

**Comment by jschwe** [comment]
>  What's the missing part to make progress on this issue?

For me, it's this part from the opening post (emphasis mine):

> I think it'd be wonderful if Cargo **automatically** set soname of cdylib to semver-major when version is >= 1.0.0 (i.e. don't set any for 0.x.x), since soname and semver-major promise basically the same thing. **However, a separate Cargo.toml field, command line flag, or env var would work as well**.

So what's missing from my point of view is a decision if cargo could automatically set the soname for `cdylib`s according to semver, or at least provide an opt-in mechanism for this, which is a bit more user-friendly than setting linker flags.

**Comment by phi-gamma** [comment]
Another argument for having Cargo set the DT_SONAME: Only Cargo knows
it is linking the final binary and not some dependency. Here’s an
example where the ``build.rs`` approach fails: https://gitlab.com/phgsng/buggy-soname
I also outlined the problem on discourse: https://users.rust-lang.org/t/89245

The gist of it is that one library that is built as a ``cdylib``
depends on another which also produces a ``cdylib``. This causes the
``build.rs`` of both libraries to be executed and Cargo ends up
appending the two extra ``link-arg``s but in the wrong order, so the
wrong SONAME ends up in the shared object. Ideally, Cargo would add
the correct SONAME argument to the linker line exactly once, when
linking the final shared object.

(This is not a contrived usecase but based on a real world product that
is unfortunately proprietary.)

**Comment by weihanglo** [comment]
See also

* https://github.com/rust-lang/cargo/issues/2273

**Comment by Fernthedev** [comment]
I ran into this issue recently when using the ELF data of my dynamic libraries to load required dependencies through topological order and found Rust `cdylib` libraries do not include the `soname` required for linking. Various users have mentioned other use cases that depend on this too.

I'm wondering if this issue has been forgotten or if the original discussion's stance still applies here. It seems to me very odd considering that this is a build type designed for a C environment rather than a Rust environment as `dylib`. The original discussion also mentions the name of the generated `soname` would matter, but I'm not so sure since perhaps a field to be set to override the default. The default would then be `lib{crate}.so` or whatever. I'd like to revive the discussion as it has been about a year since the last message.

So to be clear, are we just left using linker flags to solve this for the foreseeable future? I'll be using the helper crate https://crates.io/crates/cdylib-link-lines until Cargo provides a mechanism for this. Thanks <3

**Comment by lu-zero** [comment]
Depending on your needs you might use [cargo-c](https://crates.io/crates/cargo-c), since it usually supports more targets and provides also a mean to install the library.

**Comment by morganava** [comment]
I ended up working around this relying on cmake to do the right thing with the shared library (i.e. build the Rust crate as a shared-lib, forward the calls from static lib to shared lib symbols in a single C file; see: https://github.com/blueprint-freespeech/gosling/blob/main/source/gosling/crates/cgosling-proc-macros/src/lib.rs )

**Comment by Fernthedev** [comment]
> Depending on your needs you might use [cargo-c](https://crates.io/crates/cargo-c), since it usually supports more targets and provides also a mean to install the library.

I'm using [https://github.com/bbqsrc/cargo-ndk](cargo-ndk) so I'm not sure I'm able to use `cargo-c` for this. Thanks though! I'm sure this could be resolved in cargo itself for cdylib.

**Comment by lu-zero** [comment]
cargo-ndk uses cargo-metadata so it cannot do what cargo-c does, but ideally it could use it, we already have support for the android target.
